hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py

```diff
@@ -52,50 +53,65 @@ def left_is_better(self, left, right):
+
+            keep = keep & current_subset
+            if len(keep) == len(current_subset):
+                return True
+            if self.consider([v for i, v in enumerate(initial) if i in keep]):
+                current_subset.intersection_update(keep)
+                return True
+            return False
+
+        # We iterate over the indices in random order. This is because deletions
+        # towards the end are more likely to work, while deletions from the
+        # beginning are more likely to have higher impact. In addition there
+        # tend to be large "dead" regions where nothing can be deleted, and
+        # by proceeding in random order we don't have long gaps in those where
+        # we make no progress.
+        self.random.shuffle(indices)
+        for i in indices:
+            candidates_for_removal &= current_subset
+            if not candidates_for_removal:
+                break
+
+            # We have already processed this index, either because it was bulk
+            # removed or is the end point of a set that was.
+            if i not in candidates_for_removal:
+                continue
+
+            candidates_for_removal.discard(i)
+            if not consider_set(current_subset - {i}):
+                continue
+
+            # We have successfully removed index i. Now try removing the region
+            # surrounding i, as deletable elements are often inside large deletable
+            # regions. Once we've tried this, we remove the end points from our
+            # candidate set, as we have tried deleting them by virtue of trying to
+            # extend the deleted region to that point and failing.
+            to_right = find_integer(
+                lambda n: i + n <= len(initial)
+                and consider_set(current_subset - set(range(i, i + n)))
             )
+            candidates_for_removal.discard(i + to_right)
 
-            # If we stopped because of an element we couldn't delete, enlarge
-            # the skipped region to include it, and continue. (If we stopped
-            # because we deleted everything, the loop is about to end anyway.)
-            skipped += 1
-
-
-# This diagram shows how we use two slices to delete (k) elements from the
-# candidate region, while retaining the other candidates, and retaining all of
-# the skipped elements.
-#                              <==================
-#          candidates                skipped
-#  /^^^^^^^^^^^^^^^^^^^^^^^^^\ /^^^^^^^^^^^^^^^^^\
-# +---+---+---+---+---+---+---+---+---+---+---+---+
-# |   |   |   |   |   |   |   |   |   |   |   |   |
-# +---+---+---+---+---+---+---+---+---+---+---+---+
-#  \_________________/ \#####/ \_________________/
-#    [:candidates-k]      k       [candidates:]
-#                      <======
+            to_left = find_integer(
+                lambda n: n <= i and consider_set(current_subset - set(range(i - n, i)))
```

Comment:
Minor: I find this condition slightly easier to read as `lambda n: i - n >= 0`, so that it mirrors the expression in the `range`.